protocol/proofs/anchor-state-poison-protection.md

```diff
@@ -0,0 +1,121 @@
+a log of game results. Game results would only be promoted to anchor states if they've passed the
+air-gap window without being blacklisted.
+
+We can avoid making significant changes to the current challenger setup by using the call to the
+`tryUpdateAnchorState` function that the `FaultDisputeGame` already makes to both store the game
+result in the game array and to use a small binary search to update the anchor state. A binary
+search is likely to be acceptable here because we can bound the starting end of the search range to
+the index of the current anchor state and only search over states that finalized after the current
+anchor state was finalized. Although this can potentially miss a few eligible anchor states under
+certain circumstances, this is an acceptable trade-off as new anchor states *will* eventually be
+found after the current anchor state.
+
+### Changes to the OptimismPortal
+
+State validity rules like blacklisting and air-gap window inside of the `OptimismPortal` would be
+removed and outsourced to the `AnchorStateRegistry`. New public getter functions in the
+`AnchorStateRegistry` would provide the information that the `OptimismPortal` needs to determine
+whether a given dispute game contract can be used to finalize a withdrawal.
+
+### Changes to the FaultDisputeGame
+
+No changes to the FaultDisputeGame contract would be required.
+
+## Upgrade and Migration Path
+
+We must guarantee that the security properties of the previous system are maintained into the
+updated system. In particular, any games that have been blacklisted in the old system must also be
+blacklisted in the updated system. Additionally, it may be necessary to forward certain getter
+functions currently available on the `OptimismPortal` contract like `disputeGameBlacklist` over to
+the `AnchorStateRegistry`.
+
+## Alternatives Considered
+
+### Use OptimismPortal in AnchorStateRegistry
+
+We considered updating the `AnchorStateRegistry` to ask the `OptimismPortal` if a particular game
+has been blacklisted. A naive version of this alternative can prevent blacklisted games from being
+used for the anchor state as long as they've been blacklisted before the air-gap window starts.
+Since we do expect some games to be blacklisted *after* the air-gap window starts (but before it
+expires) this does not fully solve the problem.
+
+We can also make the changes to the `AnchorStateRegistry` described above *without* moving the
+blacklisting and air-gap logic out of the `OptimismPortal` contract. This would still fix the
+poisoning issue but would introduce additional complexity into the `AnchorStateRegistry` without
+reducing any complexity in the `OptimismPortal` contract. Would additionally spread responsibility
+for anchor state validity across multiple contracts, which isn't ideal. However, this would have
+the benefit of not needing to touch the `OptimismPortal` at all which reduces bug surface and risk.
```

Comment:
> Since we do expect some games to be blacklisted *after* the air-gap window starts (but before it
expires) this does not fully solve the problem.

Just to make sure my understanding is correct. 
This will also apply in the second case more hybrid  case with the logic into the OptimismPortal2 but also some changes into the `AnchorStateRegistry`. 
So, here we won't be able to blacklist game after the air-gap window starts, without having the anchor possibly poisoned.

Reply:
Yeah, correct. I don't really think this hybrid thing is very useful. It adds more complexity for no good reason.

Reply:
I see thanks!
Btw, completely agreed on the complexity opinion, less complex implicitly make it more safer for us.